### PR TITLE
WIP: Unpin mpl

### PR DESCRIPTION
**Pull request by bjlittle**
This PR is attempting to unpin `mpl`. As a result of unpinning `mpl` from `1.5.3` to `2.2.2` it releases the implicit dependency pin on `numpy`, which moves from `1.13.3` to `1.14.3`.

This causes royal carnage, pepper sprayed pretty much everywhere...

Can't see quite how to not make this PR wide ranging, so I'm going to adopt the strategy of fixing individual tests per commit.

The main culprit for causing so much noise (so far) is `numpy` changing the implementation of its `np.core.arrayprint._formatArray` functionality, which we tap into via `iris.util.format_array` for printing `numpy` arrays... hence the CML armageddon. I've tried best I can to replicate legacy `numpy` `1.13` array printing for `1.14`, but there are (unavoidable) minor differences that are causing us great pain. C'est la vive!

I'm going to keep drip feeding commits into this PR until I nail all the test failures... then I'm going to sit in a dark, quiet room for a while.

**Outstanding Actions:**

- [x] Fix CellMeasures tests
- [x] Fix graphic `test_plot` errors for `test_zx` and `test_tx`
- [x] Create new `citation` graphic test
- [x] ~Fix NetCDF/HDF tmp file error~
        - legitimate HDF error output from `test_netcdf.SavePermissions` tests.
- [x] Add `iris-grib` example graphic test hash
- [x] Investigate example test TestPolynomialFit mpl UserWarning
        - raised https://github.com/SciTools/iris/issues/3032 to cover this
- [x] Investigate example `test_COP_maps` usage of `linewidth`
- [x] Fix unit test axes tick labels failures
- [x] Merge companion `test-iris-imagehash` PR https://github.com/SciTools/test-iris-imagehash/pull/15

### CODE REVIEW

**Comment by pp-mo** [comment]
IMO CML is not really fit for purpose, and actually never has been.
Seriously, could we not at this point work to retire CML instead ??

**Comment by bjlittle** [comment]
@pp-mo :+1: For retiring / replacing CML, but not at the point when we're making major changes.

It would just be totally impossible to track the differences. It's hard enough as it is, so I'd say not now and not in this PR. But definitely in a separate, focused PR for that purpose.

**Comment by bjlittle** [comment]
@pp-mo Actually, `github` has changed (again) for the better, and you can now do diffs but with whitespace differences hidden, see [this](https://github.com/SciTools/iris/pull/3019/files?utf8=%E2%9C%93&diff=split&w=1) as an example...

**Comment by pp-mo** [comment]
> you can now do diffs but with whitespace differences hidden, see this as an example...

Oh yeah, great spot -- that really is progress ! :open_mouth: :joy_cat:

File: lib/iris/util.py [hunk 1]
```diff
@@ -936,16 +936,26 @@ def format_array(arr):
...
-                                           next_line_prefix='\t\t',
-                                           separator=', ', edge_items=3,
-                                           summary_insert=summary_insert)[:-1]
+    formatArray = np.core.arrayprint._formatArray
+    max_line_len = 50
+    legacy = '1.13'
+    if arr.size > summary_threshold:
+        summary_insert = '...'
+    options = np.get_printoptions()
+    options['legacy'] = legacy
```
Comment (pp-mo) [inline lib/iris/util.py]: Do we really need to do this, if virtually every CML is changing anyway ??

N.B. I have no idea how much more carnage that might be !!
So far (https://github.com/bjlittle/iris/tree/d910238b43bf3e5b57ff35c4ac813d7a077626c9/  ~= 9am Thu 17 May) we have 77 CML files changed out of 569 ???

**Comment by pp-mo** [comment]
I think you may need to rebase, there is still a check_graphic action failure :
```
FAIL: test_cube_animation (iris.tests.experimental.test_animate.IntegrationTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/scitools_iris-2.1.0.dev0-py2.7.egg/iris/tests/experimental/test_animate.py", line 82, in test_cube_animation
    self.check_graphic()
  File "/home/travis/miniconda/envs/test-environment/lib/python2.7/site-packages/scitools_iris-2.1.0.dev0-py2.7.egg/iris/tests/__init__.py", line 842, in check_graphic
    raise AssertionError(emsg.format(msg))
AssertionError: Image comparison failed: Bad phash fe81c17e817e3e81817e7e81857e7a817e81c17e7e81c17e7a81817e817e842e with hamming distance [44] for test iris.tests.experimental.test_animate.IntegrationTest.test_cube_animation.0.
-------------------- >> begin captured logging << --------------------
PIL.PngImagePlugin: DEBUG: STREAM 'IHDR' 16 13
PIL.PngImagePlugin: DEBUG: STREAM 'sBIT' 41 4
PIL.PngImagePlugin: DEBUG: 'sBIT' 41 4 (unknown)
PIL.PngImagePlugin: DEBUG: STREAM 'pHYs' 57 9
PIL.PngImagePlugin: DEBUG: STREAM 'tEXt' 78 57
PIL.PngImagePlugin: DEBUG: STREAM 'IDAT' 147 4821
--------------------- >> end captured logging << ---------------------
```

We fixed this? ...

File: lib/iris/util.py (hunk 1 above)
Comment (bjlittle) [reply lib/iris/util.py]: Yes, we really need to do this.

Comment (bjlittle) [reply lib/iris/util.py]: To put this in context... I'm down to my last 17 test failures and 1 error (excluding graphical tests), so I'm in the home straight. I don't expect another ~500 CML file changes :pray:

Comment (bjlittle) [reply lib/iris/util.py]: The graphical tests will be a different can of worms...

**Comment by pp-mo** [comment]
Just did a quick scan of the test output log.

> Do we really need to do this, if virtually every CML is changing anyway ?? ...N.B. I have no idea how much more carnage that might be !!
Status update:
Looks like we only have 2-3 CML mismatches outstanding (but in many cases more errors may emerge when those are fixed)

Then there are a bunch of about 125 image failures, most of which are reporting just "Image sizes do not match expected size"
Is that a limitation of imagehash ??

Also we currently have ~13 PP file mismatches.
I'm a bit concerned about that.

**Comment by bjlittle** [comment]
@pp-mo `mpl` `2.0` is quite different to `1.5.3`... infact it's very different. I'm expecting to see pretty much all the graphical tests fail. For a starter the `mpl` default `png` sizes are different... it would be nice if we were invariant to that. I'm not too concerned at the moment.

Let me get to the point where everything is passing apart from the graphical tests... I'm pretty close to that point. Then it'll be time to pause and work out what to do with the graphical tests.

**Comment by pp-mo** [comment]
> Let me get to the point where everything is passing apart from the graphical tests... I'm pretty close to that point. 

:+1:

File: lib/iris/tests/__init__.py [hunk 2]
```diff
@@ -1174,7 +1174,7 @@ class MyPlotTests(test.GraphicsTest):
 
 
 skip_sample_data = unittest.skipIf(not SAMPLE_DATA_AVAILABLE,
-                                   ('Test(s) require "iris_sample_data", '
+                                   ('Test(s) require "iris-sample-data", '
```
Comment (pelson) [inline lib/iris/tests/__init__.py]: 🤣

File: lib/iris/experimental/regrid.py [hunk 3]
```diff
@@ -390,7 +390,7 @@ def _weighted_mean_with_mdtol(data, weights, axis=None, mdtol=0):
             weights_sum = weights.sum(axis=axis)
             frac_masked = 1 - np.true_divide(unmasked_weights_sum, weights_sum)
             mask_pt = frac_masked > mdtol
-            if np.any(mask_pt):
```
Comment (bjlittle) [inline lib/iris/experimental/regrid.py]: A slight behaviour change between `numpy` versions `1.13.3` and `1.14.3` - you can no longer change the `mask` of a `MaskedConstant`. This makes total sense, and `numpy` have fixed quite rightly made this stricter.
```python
>>> import numpy as np
>>> np.__version__
'1.14.3'
>>> import numpy.ma as ma
>>> a = ma.array([1.0], mask=True)
>>> m = a[0]
>>> type(m)
<class 'numpy.ma.core.MaskedConstant'>
>>> m.mask |= True
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: output array is read-only
```
as opposed to
```python
>>> import numpy as np
>>> np.__version__
'1.13.3'
>>> import numpy.ma as ma
>>> a = ma.array([1.0], mask=True)
>>> m = a[0]
>>> type(m)
<class 'numpy.ma.core.MaskedConstant'>
>>> m.mask |= True
>>> m
masked
>>> m.mask |= False
>>> m
masked
```

**Comment by bjlittle** [comment]
@pp-mo and @pelson 

Okay, so I'm pretty much done here...finally :smile: 

All the tests are passing, apart from one final failure!

This simply requires the companion `test-iris-imagehash` PR https://github.com/SciTools/test-iris-imagehash/pull/15 to be merged in order to support the perceptual image hash changes check of the `imagerepo.json` in the `iris.tests.test_image_json.TestImageFile.test_resolve` test.

If you're happy to merge that PR, I'll rerun the `default` tests on `travis` for `2.7` and `3.6`.

**Comment by bjlittle** [comment]
@pp-mo and @pelson 

Holy moly... the tests are all passing :bowtie:

**Comment by pp-mo** [comment]
Plugging away at these with #1 eyeball.
 . . .

File: docs/iris/example_code/Meteorology/COP_maps.py [hunk 4]
```diff
@@ -114,7 +113,7 @@ def main():
         plt.subplot(122)
         plt.title('HadGEM2 A1B-Image Scenario',  fontsize=10)
         contour_result = iplt.contourf(delta_a1b, levels, colors=colors,
-                                       linewidth=0, extend='both')
```
Comment (pelson) [inline docs/iris/example_code/Meteorology/COP_maps.py]: Would be interested to see that this is the right thing to do. Will dig a little.

File: lib/iris/tests/test_plot.py [hunk 5]
```diff
@@ -938,5 +938,28 @@ def test_plot_tmerc(self):
         self.check_graphic()
 
 
+@tests.skip_plot
```
Comment (pelson) [inline lib/iris/tests/test_plot.py]: Hello? What's the deal here then? How come this is done in this PR?

File: lib/iris/util.py (hunk 1 above)
Comment (pelson) [reply lib/iris/util.py]: I agree with merging this as it is. I'd like to follow it up with another PR though and change all ~500 CML to modern numpy.

**Pull request merged**